### PR TITLE
Class deprecation support

### PR DESCRIPTION
**Pull request by wackoisgod**
Added the support for class level deprecation which will in turn also deprecate any fields that are currently using that type.

It was requested that support for deprecating a class in the c# code generation was added. In this implementation we deprecate the class as well as any fields that are also using that class. 

I think this ensures that users don't continue to use the class in other places, maybe some food for thought that its another option to allow for class->field level deprecation.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by wackoisgod** [comment]
I signed it!

**Comment by acozzette** [comment]
ok to test

File: src/google/protobuf/compiler/csharp/csharp_message.cc [hunk 1]
```diff
@@ -98,12 +98,21 @@ const std::vector<const FieldDescriptor*>& MessageGenerator::fields_by_number()
   return fields_by_number_;
 }
 
+void MessageGenerator::AddDeprecatedFlag(io::Printer* printer) {
+  if (descriptor_->options().deprecated())
+  {
```
Comment (acozzette) [inline src/google/protobuf/compiler/csharp/csharp_message.cc]: One small thing: our usual C++ style is to put the opening curly brace on the same line as the `if (...)`.

File: src/google/protobuf/compiler/csharp/csharp_field_base.cc [hunk 2]
```diff
@@ -126,6 +126,16 @@ void FieldGeneratorBase::AddDeprecatedFlag(io::Printer* printer) {
   {
     printer->Print("[global::System.ObsoleteAttribute]\n");
   }
+  else
```
Comment (acozzette) [inline src/google/protobuf/compiler/csharp/csharp_field_base.cc]: To cut down on the block nesting I would suggest rewriting it like this:

```
else if (descriptor_->type() == FieldDescriptor::TYPE_MESSAGE &&
         descriptor_->message_type()->options().deprecated()) {
  printer->Print("[global::System.ObsoleteAttribute]\n");
}
```

**Comment by acozzette** [comment]
@wackoisgod I made a couple of small comments but it looks great otherwise, thanks for the pull request!

**Comment by wackoisgod** [comment]
@acozzette updated based on your feedback

**Comment by acozzette** [comment]
Thanks, @wackoisgod! I'll wait for the Travis run to finish just to be safe and then merge this. I see Jenkins has an error but it looks like it's unrelated to this.

**Pull request merged**